include/revng/Pipeline/InputOutputContract.h

```diff
@@ -0,0 +1,201 @@
+    PreservedInput(PreservedInput) {}
+
+  constexpr InputOutputContract(const Kind &Target,
+                                size_t PipeArgumentTargetIndex = 0) :
+    Source(nullptr),
+    InputContract(Exactness::Exact),
+    TargetKind(&Target),
+    PipeArgumentSourceIndex(0),
+    PipeArgumentTargetIndex(PipeArgumentTargetIndex),
+    PreservedInput(false) {}
+
+  constexpr InputOutputContract(const Kind &Source,
+                                Exactness InputContract,
+                                size_t PipeArgumentSourceIndex = 0,
+                                bool PreservedInput = false) :
+    Source(&Source),
+    InputContract(InputContract),
+    TargetKind(nullptr),
+    PipeArgumentSourceIndex(PipeArgumentSourceIndex),
+    PipeArgumentTargetIndex(PipeArgumentSourceIndex),
+    PreservedInput(PreservedInput) {}
+
+  void deduceResults(ContainersStatus &StepStatus,
+                     llvm::ArrayRef<std::string> ContainerNames) const;
+
+  void deduceResults(ContainersStatus &StepStatus,
+                     TargetsList &Results,
+                     llvm::ArrayRef<std::string> ContainerNames) const;
+
+  void deduceResults(ContainersStatus &StepStatus,
+                     ContainersStatus &Results,
+                     llvm::ArrayRef<std::string> ContainerNames) const;
+
+  ContainersStatus
+  deduceRequirements(const ContainersStatus &PipeOutput,
+                     llvm::ArrayRef<std::string> ContainerNames) const;
+
+  bool forwardMatches(const ContainersStatus &Status,
+                      llvm::ArrayRef<std::string> ContainerNames) const;
+  bool backwardMatches(const ContainersStatus &Status,
+                       llvm::ArrayRef<std::string> ContainerNames) const;
+
+  void insertDefaultInput(ContainersStatus &Status,
+                          llvm::ArrayRef<std::string> ContainerNames) const;
+
+private:
+  void forward(Target &Input) const;
+  bool forwardMatches(const Target &Input) const;
+  void forwardGranularity(Target &Input) const;
+
+  ///
+  /// Target fixed -> Output must be exactly Target
+  /// Target same as Source, Source derived from base ->  Most strict between
+  /// source and target Target same as source, source exactly base -> base
+  ///
+  void backward(Target &Output) const;
+  Exactness backwardInputContract(const Target &Output) const;
+  void backwardGranularity(Target &Output) const;
+  const Kind &backwardInputKind(const Target &Output) const;
+  bool backwardMatches(const Target &Output) const;
+
+  // target is the container in which the Pipe would write when used to
+  // produce the targets
+  void deduceRequirements(TargetsList &SourceContainer,
+                          TargetsList &TargetContainer) const;
+};
+
+/// An atomic contract is the way contracts can be composed.
+/// All subcontracts inside an atomic contract are evaluated in parallel
+/// on the same input and their, and their output is merged.
+class AtomicContract {
```

Comment:
> An atomic contract is the way contracts can be composed.

Maybe *atomic* is not a good name for this. An atom is usually something that cannot be further split and in groups creates something more sophisticated.

We wanted to say something like `AtomicallyExecutedContractGroup` I guess. What about just `ContractGroup` and let the atomic part in the doc?